### ISSUE TITLE
Performance regression in TMGrammar highlighting

### ISSUE THREAD

**Issue by andrewbranch** [microsoft/vscode#154568]
Does this issue occur when all extensions are disabled?: Yes/No

Version: 1.70.0-insider
Commit: 3104db414c8fedcf6e4493f14da7df0b7413853a
Date: 2022-07-08T05:16:29.757Z
Electron: 18.3.5
Chromium: 100.0.4896.160
Node.js: 16.13.2
V8: 10.0.139.17-electron.0
OS: Darwin x64 21.6.0

I’ve noticed a regression in syntax highlighting time (I have semantic highlighting disabled) in Insiders in the last week or two, very noticeable in large files (checker.ts). Opening the 45k-line file takes about 4–6 seconds to highlight where it used to be sub-second. I thought it was strange that until highlighting completes, the color I see is the color of string literals. That made me wonder if this was due not to a change in VS Code but the introduction of a template literal string somewhere in our file that’s very expensive to parse with the TextMate grammar. I attempted to check this theory by seeing if I could inspect the token scopes while everything looks like a string, but that makes the performance of all the UI in the window incredibly slow:

https://user-images.githubusercontent.com/3277153/178074112-99fd90c5-a24b-4705-9bf3-e985d59c97ae.mp4

I can reproduce this reliably. Let me know if you’d like me to capture some logs or traces or something.

**Comment by andrewbranch** [comment]
It’s a bit hard to tell what interactions are happening in the video, but if you watch the blinking of the text insertion cursor, it hangs for a few seconds every time I click somewhere else in the document.

**Comment by alexdima** [comment]
TL;DR: A heuristic we have gets broken due to the string literal that contains content on column 1.

We have a command called `Developer: Force Retokenize` which will print the time it takes to tokenize top to bottom the current file. Looking at [version A](https://raw.githubusercontent.com/microsoft/TypeScript/28dc248e5c500c7be9a8c3a7341d303e026b023f/src/compiler/checker.ts) (with the `` ` `` literal) and [version B](https://raw.githubusercontent.com/microsoft/TypeScript/c725456d9bc3d5d73bd2e7f2336bbd7c9c2bd2b4/src/compiler/checker.ts) (where the `` ` `` literal was removed), they both take around 7150 ms to tokenize top-to-bottom. So the `` ` `` literal is not introducing a performance problem.

When opening a large file and scrolling to a position, in the time until it takes the top-down tokenizer to reach the viewport, we will do a heuristic where we will tokenize just the current viewport. When being scrolled towards the end of `checker.ts`, the viewport always begins with an indented line e.g.

<img width="746" alt="image" src="https://user-images.githubusercontent.com/5047891/179904877-e728fdf9-9d6e-47e4-95ff-e3c76ed861f2.png">

Often, an indented line is part of a deep syntactical construct, like code in a member, a class declaration, or a namespace. So using the initial tokenizer state to tokenize directly the top of the viewport might produce bad results... The heuristic we do is to walk the file backwards and pick the lines that have a smaller indentation level than the viewport start line, with the hope that those lines contain the declaration start for various constructs.

For version B, that looks like this:
```txt
namespace ts { // <-- picked up from line number 2
    namespace JsxNames { // <-- picked up from line 45595
        export const IntrinsicElements = "IntrinsicElements" as __String; // <-- first visible line
```

For version A, that looks like this:
```txt
m2: ${(this.mapper2 as unknown as DebugTypeMapper).__debugToString().split(\"\\n\").join(\"\\n    \")}`; // <-- picked up from line number 355
    namespace JsxNames { // <-- picked up from line 45595
        export const IntrinsicElements = "IntrinsicElements" as __String; // <-- first visible line
```

So that is why in version A, the viewport is incorrectly tokenized as a string until the top-down parsing reaches the viewport (roughly after 7s).

The heuristic works well in general, but not when `` ` `` multiline strings are involved. There's also nothing immediate that comes to mind, since the heuristic is language agnostic and based on indentation. Given you are unblocked by moving that string literal out, I wouldn't do anything more on our side.

**Comment by andrewbranch** [comment]
Thanks for the thorough explanation!
